### PR TITLE
Implement auto i4

### PR DESCRIPTION
**Pull request by viandoxdev**
This PR implements auto i4, and device completion detection

TODO:
 - [ ] More testing
 - [ ] Remove ugly code duplication between legit and non legit version
 - [ ] Figure out why risky tag isn't working

https://github.com/user-attachments/assets/906973a7-b301-459c-96e4-d1b65e69bd01

(not sure how I managed to fuck up commit history but whatever, wanted to try doing it from IDEA but it did this so I'll do from command line next time I guess)

### CODE REVIEW

File: odinmain/src/main/kotlin/me/odinmain/utils/skyblock/ItemUtils.kt [hunk 1]
```diff
@@ -69,6 +70,22 @@ val ItemStack?.isShortbow: Boolean
...
+    get() {
+        return this?.lore?.any { it.contains("FISHING ROD") } == true
+    }
+
+/**
+ * Returns if an item is Spirit leaps or an Infinileap
+ */
+val ItemStack?.isLeap: Boolean
+    get() {
+        return this?.unformattedName?.noControlCodes?.equalsOneOf("Infinileap", "Spirit Leap") ?: false
```
Comment (freebonsai) [inline odinmain/src/main/kotlin/me/odinmain/utils/skyblock/ItemUtils.kt:86]: use itemid

File: odin/src/main/kotlin/me/odin/features/impl/floor7/p3/ArrowsDevice.kt [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    @SubscribeEvent
+    fun onTick(tickEvent: ClientTickEvent) {
+        if(!isPlayerInRoom()) {
+            reset()
+            return
+        }
+
+        if(!isDeviceComplete && activeArmorStand?.name == ACTIVE_DEVICE_STRING)  {
```
Comment (freebonsai) [inline odin/src/main/kotlin/me/odin/features/impl/floor7/p3/ArrowsDevice.kt:115]: formatting

File: odin/src/main/kotlin/me/odin/features/impl/floor7/p3/ArrowsDevice.kt [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+            // Reset is called when leaving the device room, but device remains complete across an entire run, so this doesn't belong in reset
+            isDeviceComplete = false
+        }
+    }
+
+    private fun isPlayerOnStand(): Boolean {
+        return (mc.thePlayer?.distanceSquaredTo(standPosition.toVec3()) ?: Double.MAX_VALUE) <= 1.0
+    }
+
+    private fun isPlayerInRoom(): Boolean {
```
Comment (freebonsai) [inline odin/src/main/kotlin/me/odin/features/impl/floor7/p3/ArrowsDevice.kt:85]: make both of these val getters, just cleaner

File: odinclient/src/main/kotlin/me/odinclient/features/impl/floor7/p3/ArrowsDevice.kt [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+    // ArmorStand showing the status of the device
+    private var activeArmorStand: EntityArmorStand? = null
+
+    // Clock used for action delay (mostly for phoenix swapping and leap)
+    private val clock = Clock(delay)
+    private val actionQueue = ArrayDeque<() -> Unit>()
+
+    // Number of server ticks since the last target disappeared, or null if there is a target
+    private var serverTicksSinceLastTargetDisappeared: Int? = null;
+
```
Comment (freebonsai) [inline odinclient/src/main/kotlin/me/odinclient/features/impl/floor7/p3/ArrowsDevice.kt:76]: do all the same stuff as i commented in legit version

File: odinclient/src/main/kotlin/me/odinclient/features/impl/floor7/p3/ArrowsDevice.kt [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+    // Position of the pressure plate for auto
+    private val standPosition = BlockPos(63.5, 127.0, 35.5)
+    private val roomBoundingBox = AxisAlignedBB(20.0, 100.0, 30.0, 89.0, 151.0, 51.0)
+    private val lastGateBlock = BlockPos(8, 118, 50)
+
+    private const val ACTIVE_DEVICE_STRING = "\u00A7aDevice"
+    private const val INACTIVE_DEVICE_STRING = "\u00A7cInactive"
+
+    private val classes = listOf(DungeonClass.Archer, DungeonClass.Mage, DungeonClass.Berserk, DungeonClass.Healer, DungeonClass.Tank)
```
Comment (freebonsai) [inline odinclient/src/main/kotlin/me/odinclient/features/impl/floor7/p3/ArrowsDevice.kt:408]: DungeonClass.entries?

File: odinclient/src/main/kotlin/me/odinclient/features/impl/floor7/p3/ArrowsDevice.kt [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+        BlockPos(68, 128, 50), BlockPos(66, 128, 50), BlockPos(64, 128, 50),
+        BlockPos(68, 126, 50), BlockPos(66, 126, 50), BlockPos(64, 126, 50)
+    )
+
+    // Position of the pressure plate for auto
+    private val standPosition = BlockPos(63.5, 127.0, 35.5)
+    private val roomBoundingBox = AxisAlignedBB(20.0, 100.0, 30.0, 89.0, 151.0, 51.0)
+    private val lastGateBlock = BlockPos(8, 118, 50)
+
+    private const val ACTIVE_DEVICE_STRING = "\u00A7aDevice"
```
Comment (freebonsai) [inline odinclient/src/main/kotlin/me/odinclient/features/impl/floor7/p3/ArrowsDevice.kt:405]: cant you just write §  ?

**Review by freebonsai — CHANGES_REQUESTED** [review]
Just small stuff, do you have a recent video of how it works with everything?
Tell me when you have fixed my comments and I will look over again and merge.

**Comment by viandoxdev** [comment]
> Just small stuff, do you have a recent video of how it works with everything? Tell me when you have fixed my comments and I will look over again and merge.

The demo above is fairly recent the only differences with now are a few debug mod messages i removed and some formatting.

**Comment by freebonsai** [comment]
make a drop down menu for the settings so its more obvious what are settings for auto, look in PuzzleSolvers.kt for help

File: odin/src/main/kotlin/me/odin/features/impl/floor7/p3/ArrowsDevice.kt [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+            // Prevent the tick count from continuing and registering a false positive
+            // This does mean we could get a false negative but since the player is already
+            // dead knowing if the device is complete has less importance (no leaping)
+            serverTicksSinceLastTargetDisappeared = 11
+        }
+
+        execute(1000) {
+            if (DungeonUtils.getPhase() != M7Phases.P3) return@execute
+
+            // Cast is safe since we won't return an entity that isn't an armor stand
```
Comment (freebonsai) [inline odin/src/main/kotlin/me/odin/features/impl/floor7/p3/ArrowsDevice.kt:67]: you can just filterIsInstance<EntityArmorStand>() before your find and you will not need to cast

**Review by freebonsai — CHANGES_REQUESTED** [review]
Idk fix the one thing i commented about drop down and filterIsInstance and i will merge

**Comment by viandoxdev** [comment]
Is that good yet ? @freebonsai

**Review by freebonsai — APPROVED** [review]

**Comment by freebonsai** [comment]
Can you resolve all conflicts between yours and main branch and ping me in discord when you are done, i will merge it after.

**Pull request merged**